### DISCUSSION TITLE
Accessing Private Methods from a Protocol

### DISCUSSION THREAD

**Discussion by max-muoto** [microsoft/pyright#8010] [Q&A]
Perhaps this is the intended functionality, but noticed that you aren't able to use a private method through `self` annotated with a protocol, take this example:

```python
from typing import Protocol


class MyClass1(Protocol):
    def _private_meth(self) -> None: ...


class MyClass2:
    def public_meth(self: MyClass1) -> None:
        return self._private_meth()
```

[Pyright Playground
](https://pyright-play.net/?pyrightVersion=1.1.328&pythonVersion=3.12&strict=true&code=FDBmCcHsFsAIBcCeAHAlgOwOa1dZlx5YAFKeSAY0gBsQ6LqBDAZ2dgFlEBhJ1gRgAUpSOSrUAlAC5gsWbAAmAU1CwA%2BsnCoAbo3iLV0RfAAWA5ouqhxsALQA%2BWADlI6RZNgA6L3WAMWbTh5-ACZpOQVlWGQAVwAjalQKAyNTc0t3QN5mPmt7Jxc3GXC5cCNo8HRYNNAPdU0dPWSTAXEfNpAgA)

From my point of view, in cases where `self` is being annotated with a protocol (such as in the cases of mixin, for which it's a common strategy) it would be able to make sense to access private methods without running into warnings, but perhaps there is a reason why this is the case.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

To get the effect that you're looking for, `MyClass2` should explicitly subclass from the protocol class. 

```python
class MyClass2(MyClass1):
    def public_meth(self) -> None:
        return self._private_meth()
```
